wrappers/python/nntile/layer/batch_norm.py

```diff
@@ -0,0 +1,200 @@
+        self.x = x.value
+        self.eps = eps**0.5  # we will square it in forward (implementation constraints)
+
+        # TODO: add usage of learnable transforms
+        self.weight = weight
+        self.bias = bias
+
+        self.x_res_tm = x_res_tm
+        self.x_res = self.x_res_tm.value
+
+        self.x_normalized_copy = zeros(self.x_res.shape, dtype=type(x.value))
+        self.x_mean = x_mean
+        self.inv_std = inv_std
+        self.std = zeros(inv_std.shape, dtype=type(x.value))
+
+        self.std_tmp_3dim = zeros(self.x.shape[1:], dtype=type(x.value))
+        self.std_tmp_2dim = zeros(
+            self.x.shape[1:2] + self.x.shape[3:], dtype=type(x.value)
+        )
+
+        self.n_channels = self.x.shape[1]
+        self.numel_in_channel = math.prod(self.x.shape) / self.n_channels
+
+        self.grad = None
+
+        self.redux = False
+
+        self.tmp_buff_full = zeros(self.x.shape, dtype=type(x.value))
+
+        self.grad_x_inv_denominator = zeros(self.x.shape, dtype=type(x.value))
+
+        self.tmp_buff_channels = zeros([self.x.shape[1]], dtype=type(x.value))
+        self.tmp_buff_channels_internal = zeros([self.x.shape[1]], dtype=type(x.value))
+
+    @classmethod
+    def generate_simple(cls, x, eps: float = 1e-05, redux=False):
+        x_res = TensorMoments(
+            zeros(x.value.shape, dtype=type(x.value)),
+            zeros(x.value.shape, dtype=type(x.value)),
+            grad_required=True,
+        )
+        n_channels = x.value.shape[1]
+        x_mean = zeros([n_channels], dtype=type(x.value))
+        inv_std = zeros([n_channels], dtype=type(x.value))
+
+        weight = ones([n_channels], dtype=type(x.value))
+        bias = zeros([n_channels], dtype=type(x.value))
+
+        return BatchNorm2d(
+            x, weight, bias, x_res, x_mean, inv_std, eps=eps, redux=redux
+        )
+
+    def forward_async(self):
+        # copy tensor
+        # TODO: add add_fiber3_async and remove copy
+        add_async(1.0, self.x, 0.0, self.x_res)
+
+        # mean
+        sum_fiber_async(1.0 / self.numel_in_channel, self.x, 0.0, self.x_mean, 1, 0)
+
+        # X_res = X - mean
+        add_fiber_async(-1.0, self.x_mean, 1.0, self.x_res, 1, 0)
+        # copy for backward
+        add_async(1.0, self.x_res, 0.0, self.x_normalized_copy)
+
+        # inverse std
+        norm_slice_async(1.0, self.x_res, 0.0, self.std_tmp_3dim, 0, redux=self.redux)
+        norm_slice_async(
+            1.0, self.std_tmp_3dim, 0.0, self.std_tmp_2dim, 2, redux=self.redux
+        )
+        norm_slice_async(
+            1.0 / self.numel_in_channel**0.5,
+            self.std_tmp_2dim,
+            0.0,
+            self.std,
+            1,
+            redux=self.redux,
+        )
+
+        add_async(1.0, self.std, 0.0, self.inv_std)  # copy
+        hypot_scalar_inverse_async(self.eps, 1.0, self.inv_std)
+
+        # X_res = X_res/std
+        prod_fiber_async(self.inv_std, 1.0, self.x_res, 1)
```

Comment:
renormalization by gamma and beta is missing after this step. This is how it is done LayerNorm:
```
        # Finally, normalize input
        prod_slice_async(self.inv_stddev, 1.0, self.tmp_y_value, self.axis)
        # inv_stddev can be offloaded from GPU
        self.inv_stddev.wont_use()
        # Scale normalized input for the backward phase
        prod_fiber3_async(self.gamma.value, 1.0, self.tmp_y_value, \
                self.y.value, self.axis)
        # tmp_Y_value can be offloaded from GPU
        self.tmp_y_value.wont_use()
        # gamma can be offloaded from GPU
        self.gamma.value.wont_use()
        # Shift output
        add_fiber_async(1.0, self.beta.value, 1.0, self.y.value, self.axis, 0)
```

Reply:
done